File: VBLib_Common/AnimalContext.vb
Imports Microsoft.EntityFrameworkCore
Imports Microsoft.Extensions.Logging
Imports VBLib_Common.Model


Public Class AnimalContext
    Inherits DbContext

    Public Property Persons As DbSet(Of Person)
    Public Property Animals As DbSet(Of Animal)
    Public Property Beavers As DbSet(Of Beaver)
    Public Property Crows As DbSet(Of Crow)
    Public Property Deers As DbSet(Of Deer)
    Public Property Clubs As DbSet(Of Club)
    Public Property Grades As DbSet(Of Grade) 
    Public Property Jobs As DbSet(Of Job)
    Public Property Drawbacks As DbSet(Of Drawback)
    Public Property JobDrawbacks As DbSet(Of JobDrawback)
    Public Property Food As DbSet(Of Food)
    Public Property NormalFood As DbSet(Of NormalFood)
    Public Property VeganFood As DbSet(Of VeganFood)
    Public Property MapToQuery As DbSet(Of MapToQuery)

    Public Function GetAnimalLocation(animalId As Integer) As IQueryable(Of AnimalLocation)
        Return FromExpression(Function() GetAnimalLocation(animalId))
    End Function

    Public Sub New()
    End Sub

    Public Sub New(options As DbContextOptions(Of AnimalContext))
        MyBase.New(options)
    End Sub

    Protected Overrides Sub OnConfiguring(optionsBuilder As DbContextOptionsBuilder)
        If Not optionsBuilder.IsConfigured Then
            optionsBuilder.UseSqlServer(
                "Server=unit-1019\sqlexpress;Database=BeaversLife;Trusted_Connection=True;" &
                "MultipleActiveResultSets=True")
            optionsBuilder.LogTo(AddressOf Console.WriteLine, LogLevel.Information)
            optionsBuilder.AddInterceptors(New MySaveChangesInterceptor())
            optionsBuilder.UseLazyLoadingProxies(False)
        End If
    End Sub

    Protected Overrides Sub OnModelCreating(modelBuilder As ModelBuilder)
        modelBuilder.Entity (Of Animal)().ToTable("Animals")
        modelBuilder.Entity (Of Beaver)().ToTable("Beavers")
        modelBuilder.Entity (Of Crow)().ToTable("Crows")
        modelBuilder.Entity (Of Deer)().ToTable("Deers")
        modelBuilder.Entity (Of Animal)().HasMany(Function(a) a.Clubs).WithMany(Function(c) c.Animals).UsingEntity _
            (Of AnimalClub)(
                Function(c) _
                               c.HasOne(Function(animalClub) animalClub.Club).WithMany().HasForeignKey(
                                   Function(animalClub) animalClub.ClubId),
                Function(a) _
                               a.HasOne(Function(animalClub) animalClub.Animal).WithMany().HasForeignKey(
                                   Function(animalClub) animalClub.AnimalId), Function(j)
                                       j.[Property](Function(pt) pt.PublicationDate).HasDefaultValueSql(
                                           "CURRENT_TIMESTAMP")
                                       j.HasKey(Function(t) New With {t.AnimalId, t.ClubId
                                                   })
                                   End Function)
        modelBuilder.Entity (Of Grade)(Function(entity)
            entity.[Property](Function(e) e.TheGrade).HasColumnType("decimal(3, 2)").HasAnnotation(
                "Relational:ColumnType", "decimal(3, 2)")
        End Function)
        modelBuilder.Entity (Of JobDrawback)(Function(entity)
            entity.HasKey(Function(j) New With {j.JobId, j.DrawbackId
                             })
            entity.HasOne(Function(jd) jd.Job).WithMany(Function(j) j.JobDrawbacks).HasForeignKey(Function(x) x.JobId).
                                                OnDelete(DeleteBehavior.ClientSetNull)
            entity.HasOne(Function(jd) jd.Drawback).WithMany(Function(d) d.JobDrawbacks).HasForeignKey(
                Function(x) x.DrawbackId).OnDelete(DeleteBehavior.ClientSetNull)
        End Function)
        modelBuilder.Entity (Of MapToQuery)().ToSqlQuery(
            "    select b.Id, b.Fluffiness, b.Size, ac.ClubId as ClubId from Beavers b
                              inner join AnimalClub ac on b.Id = ac.AnimalId
                              Union all
                              select cr.Id, 2, 1, ac.ClubId as ClubId  from Crows cr
                              inner join AnimalClub ac on cr.Id = ac.AnimalId
                    ")
        modelBuilder.SharedTypeEntity (Of Dictionary(Of String, Object))("Category", Function(category)
            category.IndexerProperty (Of Integer)("Id")
            category.IndexerProperty (Of String)("Name").IsRequired()
            category.IndexerProperty (Of Integer?)("FoodId")
            category.HasOne(GetType(Food)).WithOne()
        End Function)
        modelBuilder.SharedTypeEntity (Of Dictionary(Of String, Object))("Product", Function(product)
            product.IndexerProperty (Of Integer)("Id")
            product.IndexerProperty (Of String)("Name").IsRequired()
            product.IndexerProperty (Of Integer?)("CategoryId")
            product.HasOne("Category", Nothing).WithMany()
        End Function)
        modelBuilder.Entity (Of Drawback)(Function(drawback)
            drawback.OwnsOne(Function(d) d.Consequence, Function(cons)
                cons.[Property](Function(c) c.Name).IsRequired()
            End Function)
            drawback.Navigation(Function(d) d.Consequence).IsRequired()
        End Function)
        modelBuilder.Entity(GetType(AnimalLocation)).HasNoKey()
        modelBuilder.HasDbFunction(Function() GetAnimalLocation("default"))
    End Sub
End Class

File: CoreLib_Common/AnimalContext.cs
using System;
using System.Collections.Generic;
using System.Linq;
using CoreLib_Common.Model;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;

namespace CoreLib_Common
{
    public class AnimalContext: DbContext
    {
        #region Tables

        public DbSet<Person>         Persons         { get; set; } = null!;
        public DbSet<Animal>         Animals         { get; set; } = null!;
        public DbSet<Beaver>         Beavers         { get; set; } = null!;
        public DbSet<Crow>           Crows           { get; set; } = null!;
        public DbSet<Deer>           Deers           { get; set; } = null!;
        public DbSet<Club>           Clubs           { get; set; } = null!;
        public DbSet<Grade>          Grades          { get; set; } = null!;
        public DbSet<Job>            Jobs            { get; set; } = null!;
        public DbSet<Drawback>       Drawbacks       { get; set; } = null!;
        public DbSet<JobDrawback>    JobDrawbacks    { get; set; } = null!;
        public DbSet<Food>           Food            { get; set; } = null!;
        public DbSet<NormalFood>     NormalFood      { get; set; } = null!;
        public DbSet<VeganFood>      VeganFood       { get; set; } = null!;
        public DbSet<Elf>            Elves           { get; set; } = null!;
        public DbSet<AdditionalInfo> AdditionalInfos { get; set; } = null!;

        public DbSet<MapToQuery> MapToQuery { get; set; } = null!;

        // Property bags
        public DbSet<Dictionary<string, object>> Products => Set<Dictionary<string, object>>("Product");
        public DbSet<Dictionary<string, object>> Categories => Set<Dictionary<string, object>>("Category");

        public IQueryable<AnimalLocation> GetAnimalLocation(
            int animalId)
        {
            return FromExpression(() => GetAnimalLocation(animalId));
        }

        #endregion

        #region C-ors

        public AnimalContext()
        {
        }

        public AnimalContext(DbContextOptions<AnimalContext> options) : base(options)
        {
        }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            if (!optionsBuilder.IsConfigured)
            {
                // TODO: fix connection property
                optionsBuilder.UseSqlServer("Server=unit-1019\\sqlexpress;Database=BeaversLife;Trusted_Connection=True;"+
                                            "MultipleActiveResultSets=True");
                //optionsBuilder.UseSqlServer("Server=localhost;Database=BeaversLife;Trusted_Connection=True;" +
                //                            "MultipleActiveResultSets=True"
                //    //, b=> b.UseQuerySplittingBehavior(QuerySplittingBehavior.SplitQuery)
                //);
                optionsBuilder.LogTo(Console.WriteLine, LogLevel.Information);
                optionsBuilder.AddInterceptors(new MySaveChangesInterceptor());

                // TODO: uncomment to use lazy loading.
                //optionsBuilder.UseLazyLoadingProxies();
                //optionsBuilder.UseLazyLoadingProxies(true);

                optionsBuilder.UseLazyLoadingProxies(false);
            }
        }

        #endregion

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            // TPT
            modelBuilder.Entity<Animal>().ToTable("Animals");
            modelBuilder.Entity<Beaver>().ToTable("Beavers");
            modelBuilder.Entity<Crow>().ToTable("Crows");
            modelBuilder.Entity<Deer>().ToTable("Deers");

            // Many-to-many
            modelBuilder.Entity<Animal>()
                .HasMany(a => a.Clubs)
                .WithMany(c => c.Animals)
                .UsingEntity<AnimalClub>(
                    c => c.HasOne(animalClub => animalClub.Club)
                        .WithMany().HasForeignKey(animalClub => animalClub.ClubId),
                    a => a.HasOne(animalClub => animalClub.Animal)
                        .WithMany().HasForeignKey(animalClub => animalClub.AnimalId),
                    j =>
                    {
                        j.Property(pt => pt.PublicationDate).HasDefaultValueSql("CURRENT_TIMESTAMP");
                        j.HasKey(t => new {t.AnimalId, t.ClubId});
                    }
                )
                ;

            // Many-to-many
            modelBuilder.Entity<AdditionalInfo>()
                        .HasMany(a => a.Clubs)
                        .WithMany(c => c.AdditionalInfos)
                        .UsingEntity<AdditionalInfoClub>(
                            c => c.HasOne(additionalInfoClub => additionalInfoClub.Club)
                                  .WithMany().HasForeignKey(additionalInfoClub => additionalInfoClub.ClubId),
                            a => a.HasOne(additionalInfoClub => additionalInfoClub.AdditionalInfo)
                                  .WithMany().HasForeignKey(animalClub => animalClub.AdditionalInfoId),
                            j => { j.HasKey(t => new {t.AdditionalInfoId, t.ClubId}); }
                        )
                ;

            modelBuilder.Entity<Grade>(entity =>
            {
                entity.Property(e => e.TheGrade)
                      .HasColumnType("decimal(3, 2)")
                      .HasAnnotation("Relational:ColumnType", "decimal(3, 2)");
            });

            // Many-to-many old style
            modelBuilder.Entity<JobDrawback>(entity =>
            {
                entity.HasKey(j => new
                {
                    j.JobId, j.DrawbackId
                });

                entity.HasOne(jd => jd.Job)
                      .WithMany(j => j.JobDrawbacks)
                      .HasForeignKey(x => x.JobId)
                      .OnDelete(DeleteBehavior.ClientSetNull);

                entity.HasOne(jd => jd.Drawback)
                      .WithMany(d => d.JobDrawbacks)
                      .HasForeignKey(x => x.DrawbackId)
                      .OnDelete(DeleteBehavior.ClientSetNull);
            });
            
            //Map entity types to queries
            modelBuilder.Entity<MapToQuery>().ToSqlQuery(
                @"    select b.Id, b.Fluffiness, b.Size, ac.ClubId as ClubId from Beavers b
                              inner join AnimalClub ac on b.Id = ac.AnimalId
                              Union all
                              select cr.Id, 2, 1, ac.ClubId as ClubId  from Crows cr
                              inner join AnimalClub ac on cr.Id = ac.AnimalId
                    ");

            // Property bags
            modelBuilder.SharedTypeEntity<Dictionary<string, object>>("Category", category =>
            {
                category.IndexerProperty<int>("Id");
                category.IndexerProperty<string>("Name").IsRequired();
                category.IndexerProperty<int?>("FoodId");

                category.HasOne(typeof(Food)).WithOne() /*.HasForeignKey("Category", "FoodId")*/;
            });

            modelBuilder.SharedTypeEntity<Dictionary<string, object>>("Product", product =>
            {
                product.IndexerProperty<int>("Id");
                product.IndexerProperty<string>("Name").IsRequired();
                product.IndexerProperty<int?>("CategoryId");

                product.HasOne("Category", null).WithMany();
            });

            // Required 1:1 dependents
            modelBuilder.Entity<Drawback>(drawback =>
            {
                drawback.OwnsOne(d => d.Consequence,
                    cons => { cons.Property(c => c.Name).IsRequired(); });
                drawback.Navigation(d => d.Consequence).IsRequired();
            });

            // Table-valued functions
            modelBuilder.Entity(typeof(AnimalLocation)).HasNoKey();
            modelBuilder.HasDbFunction(() => GetAnimalLocation(default));
        }
    }
}
